core/state_processor.go

```diff
@@ -97,6 +98,14 @@ func applyTransaction(msg types.Message, config *params.ChainConfig, bc ChainCon
 	txContext := NewEVMTxContext(msg)
 	evm.Reset(txContext, statedb)
 
+	// Compute the fee related information that is to be included
+	// on the receipt. This must happen before the state transition
+	// to ensure that the correct information is used.
+	l1Fee, _, _, err := fees.CalculateFees(tx, statedb)
```

Comment:
Isn't this calculated somewhere else already? Would it be possible to calculate it only once per transaction?

Reply:
we call VerifyFee, which also calculates fees, when transactions are added to the tx pool. however, fees could change between when they are added to the pool and when they are executed. 

we could also consider calculating fees once before any transactions for the block are executed. i think this is not 100% correct since technically a transaction that is executed after an oracle update would have a different price, but it might be a better user experience and less compute.

Reply:
I meant `CalculateL1MsgFee` which is called in `NewStateTransition`. Not sure if these two calls can be combined into one easily.

Reply:
As you mentioned, the L1 fee is calculated once when a transaction is added to the pool, and the pool discards transactions that are under-calculated based on the L1 base fee at that time.

Does the recalculation here imply that some of the transactions that were discarded at that time actually met the gas conditions at the time of execution? [here](https://github.com/scroll-tech/go-ethereum/blob/scroll/core/tx_pool.go#L685-L691)

Reply:
Discarded transactions will never get to execute. But it's possible the other way around: Transaction is added to txpool but by the time it's executed the L1 gas oracle is updated and the tx fails with out of gas.

Reply:
Ye, or a transaction that would have been discarded might actually meet the L1 Fee requirement in the future when the L1 Fee declines. Would it be better not to discard all of them when the Transaction Pool verifies the L1 Fee, but to keep some of the transactions that are close to the L1 Fee requirements for a while?